diff/walking/differ.go

```diff
@@ -56,6 +57,38 @@ func NewWalkingDiff(store content.Store) diff.Comparer {
 	}
 }
 
+func ReadonlyMounts(mounts []mount.Mount) []mount.Mount {
+	for i, m := range mounts {
+		if m.Type == "overlay" {
+			mounts[i].Options = readonlyOverlay(m.Options)
+		}
+	}
+	return mounts
+}
+
+// readonlyOverlay modifies overlay mount options to make them readonly by
+// removing workdir and upperdir (and appending the upperdir layer to lowerdir) - see:
+// https://www.kernel.org/doc/html/latest/filesystems/overlayfs.html#multiple-lower-layers
```

Comment:
Why not append `ro`?

Reply:
I didn't add it because what really matters here for `overlayfs` to prevent it from entering a bad state is to mount without the `upperdir` and `workdir`. Happy to include appending `ro` though.

Reply:
Will non-overlayfs mounts be read-only without `ro` being set here? (in that case, I definitely agree that we probably would need to add it, otherwise it will be unexpected that the mount isn't actually read-only)

Reply:
I took a page out of https://github.com/moby/buildkit/blob/3318657ae929a01e59df0cec58f49ea559938e2f/cache/refs.go#L1516 and changed the implementation to more closely match. `readonlyOverlay` has the overlay specific code, but we're appending `ro` in `readonlyMounts` now.

Reply:
Why doesn't `cache/refs.go` append `ro`?

Reply:
It does – https://github.com/moby/buildkit/blob/3318657ae929a01e59df0cec58f49ea559938e2f/cache/refs.go#L1527 – and now this implementation more closely matches that.

Reply:
Oh yes, not for overlay mounts. It doesn't hurt to include `ro` for overlay mounts I believe, but the `upperdir`/`workdir` change is what's relevant there:

For example, from https://www.kernel.org/doc/html/next/filesystems/overlayfs.html#multiple-lower-layers:

> `# mount -t overlay overlay -olowerdir=/lower1:/lower2:/lower3 /merged`
> As the example shows, “upperdir=” and “workdir=” may be omitted. In that case the overlay will be read-only.

Or from https://wiki.archlinux.org/title/Overlay_filesystem:

> Sometimes, it is only desired to create a read-only view of the combination of two or more directories. In that case, it can be created in an easier manner, as the directories upper and work are not required:
> `# mount -t overlay overlay -o lowerdir=/lower1:/lower2 /merged`

None of those examples use `ro` because it's not relevant for overlay mounts in that case.

Reply:
maybe @tonistiigi can add more information as to why buildkit doesn't add `ro`

Reply:
Just that it is not needed, as you explained already:

```
/tmp/a # mkdir lower1 lower2 merged work
/tmp/a # touch lower1/1
/tmp/a # touch lower2/2
/tmp/a # mount -t overlay overlay -o lowerdir=lower2:lower1,workdir=work merged
/tmp/a # ls -l merged/
total 0
-rw-r--r--    1 root     root             0 Mar 16 18:03 1
-rw-r--r--    1 root     root             0 Mar 16 18:03 2
/tmp/a # touch merged/3
touch: merged/3: Read-only file system
```